skbio/diversity/beta/_fast_unifrac.py

```diff
@@ -0,0 +1,390 @@
+    u_counts = count_array[:, 0]
+    v_counts = count_array[:, 1]
+
+    # fetch the lengths
+    length = indexed['length']
+
+    u = w_unifrac(length, u_counts, v_counts, u_sum, v_sum)
+    if normalized:
+        # get the index positions for tips in counts_array, and determine the
+        # tip distances to the root
+        tip_indices = np.array([n.id for n in indexed['id_index'].values()
+                                if n.is_tip()])
+        tip_ds = tip_distances(length, tree, tip_indices)
+        u /= _branch_correct(tip_ds, u_counts, v_counts, u_sum, v_sum)
+
+    return u
+
+
+def _boundary_case(u_sum, v_sum, normalized=False, unweighted=True):
+    """Test for boundary conditions
+
+    Parameters
+    ----------
+    u_sum, v_sum: float
+        The sum of the observations in both environments.
+    normalized: bool
+        Indicates if the method is normalized.
+    unweighted: bool
+        Indicates if the method is weighted.
+
+    Returns
+    -------
+    float or None
+        Specifically, one of `[0.0, 1.0, None]`. `None` indicates that a
+        boundary condition was not observed.
+
+    Notes
+    -----
+    The following boundary conditions are tested:
+
+        * if u_sum or v_sum are zero
+        * if both u_sum and v_sum are zero
+        * both of the above conditions with respect to normalized and weighted
+    """
+    if u_sum and v_sum:
+        return None
+
+    if u_sum + v_sum:
+        # u or v counts are all zeros
+        if unweighted or normalized:
+            # u or v counts are all zeros
+            return 1.0
+        # NOTE: we cannot handle the unnormalized case here yet as it
+        # requires operations on the tree vector
+    else:
+        # u and v are zero
+        return 0.0
+
+    return None
+
+
+def _branch_correct(tip_dists, i, j, i_sum, j_sum):
+    """Calculates weighted unifrac branch length correction.
+
+    Parameters
+    ----------
+    tip_dists : np.ndarray
+        1D column vector of branch lengths in post order form. Only tips
+        should be non-zero as this represents the distance from each tip to
+        root.
+    i, j : np.ndarray
+        Aggregated environment counts. This vector is expected to be in index
+        order with tip_dists
+    i_sum, j_sum: float
+        Counts of the observations in each environment
+
+    Returns
+    -------
+    np.ndarray
+        The corrected branch lengths
+    """
+    return (tip_dists.ravel() * ((i / i_sum) + (j / j_sum))).sum()
+
+
+# TODO: define a block pdist method such that "blocks" of the resulting
+# distance matrix can be computed independently. This would allow for an
+# embarassingly parallel strategy for computing large distance matrices. This
+# method needs to be paired with a DistanceMatrix assembly method, however,
+# it may be beneficial to "assemble" directly into a memory mapped data
+# structure.
+
+
+def block_pdist(observation_data):
```

Comment:
It would be nice to get this in for 0.4.1, but I think it should be lowest on the priority list (since it relies on everything else being ready). We might also want to have our own scikit-bio ``pdist`` function that incorporates this functionality so it's easy for us to use for any of the metrics.